New SQLite driver: capacitor-sqlite

gregzo opened the pull request:
This PR adds compatibility with [capacitor-community/sqlite](https://github.com/capacitor-community/sqlite) via a new adapter and accompanying plumbing and tests.

We could envisage asking `capacitor-community/sqlite`'s maintainer to have a quick look to ensure that there's no flagrant misuse of its APIs. As it stands, all tests pass but using a mock database. 

This adapter implementation is built on my understanding of `electric` and `capacitor-community/sqlite` documentation and codebase. Much more extensive real-world testing is imo required.

kevin-dp commented on clients/typescript/src/drivers/capacitor-sqlite/database.ts:
```diff
@@ -0,0 +1,17 @@
+
+// A bit of a hack, but that lets us reference the actual types of the library
+// TODO: Is this the type we want to expose?
+type OriginalDatabase = SQLiteDBConnection;
```
We don't use semicolons.
You can automatically format your code with:
```
# CAREFUL: only run the below command within the clients/typescript folder
npx prettier --write .
```

alco replied:
We also have an npm script for formatting:
```
cd clients/typescript
pnpm format
```

gregzo replied:
Fixed by bdf9259fb6eddda5f3346a7bdbc806e54c761b8f

kevin-dp commented on clients/typescript/src/drivers/capacitor-sqlite/mock.ts:
```diff
@@ -0,0 +1,54 @@
+  dbname: DbName;
+  fail: Error | undefined;
+
+  constructor(dbName: DbName, fail?: Error) {
```
Btw, you could define those public fields on one line:

```ts
constructor(public dbName: DbName, public fail?: Error) {}
```

gregzo replied:
Done in 05cafe0a42a0544c14840f662f0da670d2aee920

kevin-dp commented on clients/typescript/src/drivers/capacitor-sqlite/mock.ts:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  executeSet(): Promise<capSQLiteChanges> {
+    return new Promise<capSQLiteChanges>( (resolve, reject) => {
```
Would this be more readable?

```ts
if (typeof this.fail !== 'undefined')
  return Promise.reject(this.fail)
else
  return Promise.resolve({ changes: { changes: 0 } })
```

gregzo replied:
Done in 05cafe0a42a0544c14840f662f0da670d2aee920

kevin-dp commented on clients/typescript/src/drivers/capacitor-sqlite/mock.ts:
```diff
@@ -0,0 +1,54 @@
+    }); 
+  }
+  query(): Promise<DBSQLiteValues> {
+    return new Promise<DBSQLiteValues>( (resolve, reject) => {
```
The pattern is always the same, reject if `this.fail` is defined, otherwise resolve to some value. We should extract this in a function:

```ts
private resolveIfNotFail<T>(value: T): Promise<T> {
  if (typeof this.fail !== 'undefined')
    return Promise.reject(this.fail)
  else 
    return Promise.resolve(value)
}
```

Then we can replace all duplicated code by a call to `resolveIfNotFail`.

gregzo replied:
Done in 05cafe0a42a0544c14840f662f0da670d2aee920

kevin-dp reviewed (changes requested):
Thanks again for your contribution!
I reviewed your PR and mostly have minor comments to improve readability. However, i do have one major concern. The `transaction<T>` method of the adapter is managing the transaction manually by calling `BEGIN` and `COMMIT`/`ABORT`.

This is prone to concurrency issues if one would call `transaction<T>` before the previous transaction finished; in which case the transactions will get interleaved. Even if we assume that clients won't do that, a similar issue could still occur when clients call `transaction<T>` (e.g. as a result of calling `db.items.create`) and concurrently there is a transaction coming in via the replication stream and being applied.

We've had similar issues with the `wa-sqlite` driver and solved them by using a mutex as to ensure that the adapter never tries to run two transactions concurrently. I suggest you have a look at the implementation of `transaction<T>` in the wa-sqlite driver.

gregzo commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,126 @@
+
+    const wrapInTransaction = false; // Default is true. electric calls run from within transaction<T> so we need to disable transactions here.
+
+    return this.db.run(sql,args,wrapInTransaction).then((result: capSQLiteChanges) => {
```
Fixed by ea396770e421e514317074525c9d77f7a5443436

gregzo commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,126 @@
+    const set: capSQLiteSet[] = statements.map( ({sql, args}) => ({statement: sql, values: args as SqlValue[] }));
+    const wrapInTransaction = true;
+
+    return this.db.executeSet(set, wrapInTransaction).then( (result: capSQLiteChanges) => {
```
Fixed by ea396770e421e514317074525c9d77f7a5443436

gregzo commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,126 @@
+      )
+    }
+
+    return this.db.query(sql, args).then( (result) => {
```
Fixed by ea396770e421e514317074525c9d77f7a5443436

gregzo commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,126 @@
+  ): Promise<T> {
+
+    return new Promise<T>( (resolve,reject) => {
+      this.db.beginTransaction().then( () => {
```
Implemented in 05cafe0a42a0544c14840f662f0da670d2aee920

gregzo commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,126 @@
+}
+
+// Did consider handling begin/commit/rollback transaction in this wrapper, but in the end it made more sense
+// to do so within the transaction<T> implementation, promises bublle up naturally that way and no need for inTransaction flag.
```
Fixed in 05cafe0a42a0544c14840f662f0da670d2aee920

gregzo commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,126 @@
+      );
+    }
+
+    const set: capSQLiteSet[] = statements.map( ({sql, args}) => ({statement: sql, values: args as SqlValue[] }));
```
Done in 05cafe0a42a0544c14840f662f0da670d2aee920

gregzo commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,126 @@
+    const wrapInTransaction = true;
+
+    return this.db.executeSet(set, wrapInTransaction).then( (result: capSQLiteChanges) => {
+      // TODO: unsure how capacitor-sqlite populates the changes value (additive?), and what is expected of electric here.
```
I'd need to write a test against a real db, not the mock. What I can do for now is ask @samwillis to double check this in his example project.

gregzo commented:
All requested changes implemented, all tests still pass. Thank you for the review! The only change that is more than a stylistic or readability refactor is the mutex implementation. I've left that conversation unresolved.

samwillis commented on clients/typescript/src/drivers/capacitor-sqlite/adapter.ts:
```diff
@@ -0,0 +1,153 @@
+
+    const set: capSQLiteSet[] = statements.map(({ sql, args }) => ({
+      statement: sql,
+      values: args as SqlValue[] | undefined,
```
Capacitor-sqlite expects an empty array when there are no bound values:

```suggestion
      values: args as SqlValue[] || [],
```

kevin-dp replied:
> Capacitor-sqlite expects an empty array when there are no bound values:

No need to add the empty array type, `SqlValue[]` also encapsulates the empty array :-)

kevin-dp replied:
I believe the issue is deeper here. capacitor-sqlite expects an empty array when there are no bind parameters, so we should not cast it to the desired type but we should instead make sure it is never undefined. So:
`values: args ?? []`

kevin-dp replied:
Ok, i see, we can cast it then 👍

gregzo replied:
Fixed in 8a301ac51d9b81eb828f77827fda5a2ee67c3e66

kevin-dp replied:
@samwillis sorry i read your comment in a hurry and clearly was confused. You are completely right.

samwillis commented:
Hey @gregzo, good news, with the two small changes I've comment on this worked with my demo app (I've tested iOS but not android yet, but I don't think that will change anything) 😁 

It looks like you also have an issue with `pnpm-lock.yaml` not being top to date (see the failed CI tests), I think all you need to do is run `pnpm install` to update it.

gregzo commented:
@samwillis pnpm-lock now up to date. FYI ran into this known pnpm issue: https://github.com/pnpm/pnpm/issues/1801
Manually created missing folders and js files to work around it.

kevin-dp reviewed (approved):
Good job @gregzo!
Driver looks great, i'm waiting for the tests to finish and will merge afterwards.
Thanks again for your amazing contribution 💯

kevin-dp commented:
@gregzo Could you run `yarn format` in `clients/typescript` please?

gregzo commented:
@kevin-dp yes of course, done in a09245e9914b8117daf0b7c8895872b122ebe2fc
My bad.

alco commented on clients/typescript/CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+
+### Patch Changes
+
+- 226c004: Fix bug for table named "model"
```
How did this change get here? All of our pending changesets are kept in the `.changeset` directory. Only when we publish a new release do they get removes and the appropriate CHANGELOG.md files are updated automatically. We never do it by hand.

alco commented on clients/typescript/package.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "version": "0.6.4",
+  "version": "0.6.5",
```
This version is also bumped automatically by our release management tool.

Add a new changeset by running `pnpm changeset` in the repo root, pick the `electric-sql` project and choose the `minor` level for the change.

alco commented on clients/typescript/src/version/index.ts:
```diff
@@ -1 +1 @@
-export const LIB_VERSION = '0.6.4'
+export const LIB_VERSION = '0.6.5'
```
As explained above, this shouldn't have been changed by hand.

alco reviewed (commented):
I see you merged `main` into your PR branch. We don't do that normally and instead rebase PR branches on top of main.

Please rebase your branch on top of the latest `main` and remove the merge commit and the Version Packages one. Let me know if you need a hand with that.

alco commented:
The E2E test failures seem to be caused by the fact that your branch is based on an older version of the server. When you rebase it on top of the latest main, the E2E problems should be resolved.

gregzo commented:
@alco Understood. The easiest would be to reset to 8a301ac51d9b81eb828f77827fda5a2ee67c3e66 and force push. After that I can rebase on main head. I that ok with you?

alco commented:
@gregzo Rebase looks great. Thanks!

I was just about to comment about the pnpm lockfile update. I've pushed a commit for it [here](https://github.com/electric-sql/electric/commit/499ecea6c92ecc5b15078ff7993c22a136085a9a).

Did you also get these warnings when you ran `pnpm install`?
```
$ pnpm install
Scope: all 6 workspace projects
clients/typescript                       |  WARN  deprecated @types/react-native-uuid@2.0.0
clients/typescript                       |  WARN  deprecated squel@5.13.0
e2e/satellite_client                     |  WARN  deprecated tslint@6.1.3
e2e/satellite_client                     |  WARN  deprecated @tsmodule/tsmodule@40.19.0
Downloading registry.npmjs.org/typescript/5.1.3: 7.15 MB/7.15 MB, done
Downloading registry.npmjs.org/typescript/4.9.5: 11.62 MB/11.62 MB, done
Downloading registry.npmjs.org/react-native/0.68.0: 21.14 MB/21.14 MB, done
Downloading registry.npmjs.org/hermes-engine/0.11.0: 15.39 MB/15.39 MB, done
Downloading registry.npmjs.org/sql.js/1.8.0: 7.45 MB/7.45 MB, done
Downloading registry.npmjs.org/@stencil/core/4.4.0: 8.75 MB/8.75 MB, done
Downloading registry.npmjs.org/jsc-android/250230.2.1: 36.09 MB/36.09 MB, done
Downloading registry.npmjs.org/dprint-node/1.0.7: 10.55 MB/10.55 MB, done
 WARN  14 deprecated subdependencies found: @babel/plugin-proposal-async-generator-functions@7.20.7, @babel/plugin-proposal-private-property-in-object@7.21.0, @babel/plugin-proposal-unicode-property-regex@7.18.6, @npmcli/move-file@1.1.2, har-validator@5.1.5, request-promise-native@1.0.9, request@2.88.2, resolve-url@0.2.1, source-map-resolve@0.5.3, source-map-url@0.4.1, uglify-es@3.3.9, urix@0.1.0, uuid@3.4.0, w3c-hr-time@1.0.2
Packages: +12
++++++++++++
Progress: resolved 1844, reused 0, downloaded 1759, added 12, done
 WARN  Issues with peer dependencies found
clients/typescript
└─┬ react-native 0.68.0
  ├── ✕ unmet peer react@17.0.2: found 18.2.0
  └─┬ react-shallow-renderer 16.14.1
    └── ✕ unmet peer react@"^16.0.0 || ^17.0.0": found 18.2.0

Done in 34.6s
```

I'm wondering if there are any fixes we need to apply to our deps. And it's curious that the lockfile version dropped to `6.0` from `6.1`, even though I had just updated my `pnpm` to the latest version yesterday.

@balegas @kevin-dp

kevin-dp commented on .changeset/odd-falcons-float.md:
```diff
@@ -0,0 +1,5 @@
+"electric-sql": minor
+---
+
+New adapter for capacitor-sqlite.
```
I think "DB driver" would be more precise than "adapter".
I'm nitpicking here but these logs will be shown when users browse the changes of the next release so "DB driver" may speak more to them than "adapter".

gregzo replied:
Sure, done.

gregzo commented:
> The E2E test failures seem to be caused by the fact that your branch is based on an older version of the server. When you rebase it on top of the latest main, the E2E problems should be resolved.

Rebased on electric/main and added changeset. E2E tests still fail though - happy to follow your lead if there's anything else I can do.

alco reviewed (approved):
:medal_sports:

gregzo commented:
@alco Yes, I encountered the same warnings.
